File: frontend/lib/sanitizar-html.ts
/**
 * Sanitizador de HTML de conteúdo editorial, sem dependência externa.
 *
 * -------------------------------------------------------------------------
 * POR QUE NÃO USAR `dangerouslySetInnerHTML` COM HTML DE RSS
 * -------------------------------------------------------------------------
 * `resumo` de item do feed é texto puro do XML da fonte, mas `feedparser`
 * também aceita `<content:encoded>`/`<description>` com HTML completo. Se
 * esse HTML chegar cru ao `dangerouslySetInnerHTML`, é XSS armazenado
 * servido sob a origem do portal.
 *
 * -------------------------------------------------------------------------
 * POR QUE ESTE MÓDULO E NÃO `sanitize-html` / `DOMPurify`
 * -------------------------------------------------------------------------
 * O projeto roda em Node 20 e 22 (CI e Dockerfiles) sem jsdom, e
 * DOMPurify precisa de DOM. `sanitize-html` é a opção robusta, mas
 * adicionaria dependência de runtime a um portal que hoje tem zero
 * sanitizador. Este sanitizador é deliberadamente CONSERVADOR: em vez de
 * tentar列举 e bloquear tags perigosas (blacklist, sempre incompleta),
 * ele processa o HTML token a token e reconstrói a saída a partir de uma
 * ALLOWLIST. O que não está na allowlist é descartado — o default é
 * "não renderizar", não "não bloquear".
 *
 * Limite conhecido e aceito: a tokenização é por regex, não por parser
 * HTML completo (diferente de `html.parser` do Python, usado no lado do
 * backend com garantia equivalente). Ela foi desenhada para NÃO ser
 * enganável pelas construções de bypass clássicas (ver testes em
 * `testes/sanitizar-html.test.mjs`), mas um parser de verdade é
 * estritamente superior. Registrado como dívida, não escondido.
 *
 * A sanitização no backend (`backend/config/sanitizar_html.py`) é a
 * fronteira primária porque roda antes da persistência; esta é a segunda
 * camada, para o HTML que já está no banco ou chega por cache.
 */

/** Tags permitidas e o que fazer com cada uma. */
const TAGS_PERMITIDAS: Record<string, "fechar" | "ignorar-conteudo"> = {
  p: "fechar",
  br: "fechar",
  hr: "fechar",
  strong: "fechar",
  b: "fechar",
  em: "fechar",
  i: "fechar",
  u: "fechar",
  s: "fechar",
  sub: "fechar",
  sup: "fechar",
  h1: "fechar",
  h2: "fechar",
  h3: "fechar",
  h4: "fechar",
  h5: "fechar",
  h6: "fechar",
  ul: "fechar",
  ol: "fechar",
  li: "fechar",
  dl: "fechar",
  dt: "fechar",
  dd: "fechar",
  a: "fechar",
  blockquote: "fechar",
  pre: "fechar",
  code: "fechar",
  // Tags cujo conteúdo é INSEGURO no contexto do elemento: mantém o texto
  // legível, descarta a marcação. É o comportamento correto para
  // `<script>alert(1)</script>` — o leitor vê "alert(1)" como texto.
  script: "ignorar-conteudo",
  style: "ignorar-conteudo",
  iframe: "ignorar-conteudo",
  object: "ignorar-conteudo",
  embed: "ignorar-conteudo",
  template: "ignorar-conteudo",
  noscript: "ignorar-conteudo",
  svg: "ignorar-conteudo",
  math: "ignorar-conteudo",
};

/** Atributos permitidos por tag. */
const ATRIBUTOS_PERMITIDOS: Record<string, string[]> = {
  "*": ["title", "lang", "dir"],
  a: ["href", "title", "rel", "target", "lang", "dir"],
  ol: ["start", "type"],
  li: ["value"],
};

/** Esquemas de URL tolerados em `href`. */
const ESQUEMAS_HREF = new Set(["http:", "https:", "mailto:"]);

/** Atributos da allowlist que são boolean HTML legítimos (sem valor). */
const ATRIBUTOS_FLAG_BOOLEANOS = new Set(["hidden"]);

/** Remove o conteúdo (e as tags) de tudo que estiver dentro de uma tag
 *  marcada como "ignorar-conteudo". Faz isso ANTES da tokenização normal,
 *  porque `<script>var x = "</scr" + "ipt>"` é o bypass clássico. */
function descartarConteudoPerigoso(html: string): string {
  let saida = html;
  for (const tag of Object.keys(TAGS_PERMITIDAS)) {
    if (TAGS_PERMITIDAS[tag] !== "ignorar-conteudo") continue;
    // `[\s\S]*?` para atravessar novaslines; case-insensitive.
    const abertura = new RegExp(`<${tag}\\b[^>]*>`, "gi");
    const fechoSimples = new RegExp(`</${tag}\\s*>`, "gi");
    // 1) Tags sem fecho: remove da abertura até o fim.
    saida = saida.replace(abertura, (match, offset: number) => {
      // Se existe um fecho depois, deixa o par para o passo 2.
      const resto = saida.slice(offset + match.length);
      return fechoSimples.test(resto) ? "" : match;
    });
    // 2) Par completo: descarta o miolo.
    saida = saida.replace(
      new RegExp(`<${tag}\\b[^>]*>[\\s\\S]*?(?:</${tag}\\s*>|$)`, "gi"),
      "",
    );
  }
  return saida;
}

function escaparTexto(texto: string): string {
  return (
    texto
      .replace(/&(?![a-zA-Z][a-zA-Z0-9]{1,10};|#[0-9]{1,7};|#x[0-9a-fA-F]{1,6};)/g, "&amp;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;")
      // Aspas e apóstrofos são inofensivos em text node, mas esta saída
      // pode ser embutida num atributo por um chamando futuro; escapá-los
      // torna o texto seguro nos DOIS contextos. Custo zero.
      .replace(/"/g, "&quot;")
      .replace(/'/g, "&#x27;")
  );
}

function escaparAtributo(valor: string): string {
  return escaparTexto(valor).replace(/"/g, "&quot;").replace(/'/g, "&#x27;");
}

/** Valida e normaliza um `href`; devolve `null` se o esquema não for permitido. */
function hrefSeguro(valorBruto: string): string | null {
  // Remove controles C0 e DEL ANTES de parsear. É o que neutraliza o
  // bypass `java\tscript:alert(1)` — o browser ignora tab/CR/LF
  // dentro de uma URL, então só filtrar por substring não bastaria.
  // ESPAÇO (0x20) NÃO é removido de propósito: `new URL` o codifica
  // como %20, que é o comportamento correto (`java script:` vira URL
  // inválida e é rejeitada logo abaixo).
  const bruto = valorBruto.trim().replace(/[\u0000-\u001f\u007f]/g, "");
  if (!bruto) return null;
  let analisado: URL;
  try {
    analisado = new URL(bruto);
  } catch {
    // Relativas (ex.: "/politica") são resolvidas contra o site e inócuas.
    if (bruto.startsWith("/") && !bruto.startsWith("//")) return bruto;
    return null;
  }
  const protocolo = analisado.protocol.toLowerCase();
  if (!ESQUEMAS_HREF.has(protocolo)) return null;
  if (protocolo !== "mailto:" && (analisado.username || analisado.password)) return null;
  return analisado.toString();
}

/**
 * Sanitiza HTML editorial, devolvendo só marcação da allowlist.
 * Texto é sempre preservado (com escape); marcação fora da allowlist é
 * removida.
 */
export function sanitizarHtmlEditorial(entrada: unknown): string {
  if (typeof entrada !== "string" || !entrada) return "";

  // 1. Neutraliza blocos de conteúdo perigoso antes de qualquer outra coisa.
  const semPerigos = descartarConteudoPerigoso(entrada);

  const TAG_OU_COMENTARIO = /<!--[\s\S]*?-->|<!\[CDATA\[[\s\S]*?\]\]>|<!\w[^>]*>|<\/?([a-zA-Z][a-zA-Z0-9-]*)((?:\s+[^>]*?)?)\/?>/g;

  let resultado = "";
  let cursor = 0;
  let match: RegExpExecArray | null;

  while ((match = TAG_OU_COMENTARIO.exec(semPerigos)) !== null) {
    resultado += escaparTexto(semPerigos.slice(cursor, match.index));
    cursor = match.index + match[0].length;

    // Comentário, CDATA ou declaration: descartados por inteiro.
    if (match[0].startsWith("<!") || match[0].startsWith("<!--")) continue;

    const nomeCru = (match[1] ?? "").toLowerCase();
    const ehFechamento = match[0].startsWith("</");
    const ehAutoFechada = match[0].endsWith("/>");

    // Tag não permitida (ex.: `img`, `form`, `marquee`): descarta a tag,
    // preserva o texto interno.
    if (!(nomeCru in TAGS_PERMITIDAS)) continue;
    if (TAGS_PERMITIDAS[nomeCru] === "ignorar-conteudo") continue;

    const atributos = ehFechamento ? {} : extrairAtributos(match[2] ?? "");
    const atributosLimpados = filtrarAtributos(nomeCru, atributos, ehAutoFechada);

    if (ehFechamento) {
      resultado += `</${nomeCru}>`;
    } else {
      const sufixo = ehAutoFechada ? " /" : "";
      const serializados = atributosLimpados
        .map(([chave, valor]) => (valor === null ? ` ${chave}` : ` ${chave}="${valor}"`))
        .join("");
      resultado += `<${nomeCru}${serializados}${sufixo}>`;
    }
  }
  resultado += escaparTexto(semPerigos.slice(cursor));
  return resultado;
}

/** Extrai `nome="valor"`, `nome='valor'`, `nome=valor` e flags soltas. */
function extrairAtributos(bruto: string): Record<string, string | null> {
  const saida: Record<string, string | null> = {};
  const REGEX = /([a-zA-Z_:][a-zA-Z0-9_.:-]*)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s"'>=`]+)))?/g;
  let m: RegExpExecArray | null;
  while ((m = REGEX.exec(bruto)) !== null) {
    const nome = (m[1] ?? "").toLowerCase();
    if (nome in saida) continue; // duplicata: vence a primeira
    const valor = m[2] ?? m[3] ?? m[4];
    saida[nome] = valor === undefined ? null : valor;
  }
  return saida;
}

function filtrarAtributos(
  tag: string,
  atributos: Record<string, string | null>,
  ehAutoFechada: boolean,
): Array<[string, string | null]> {
  const permitidosGlobais = ATRIBUTOS_PERMITIDOS["*"] ?? [];
  const permitidosDaTag = ATRIBUTOS_PERMITIDOS[tag] ?? [];
  const saida: Array<[string, string | null]> = [];

  for (const [nome, valor] of Object.entries(atributos)) {
    // `on*` (handlers) nunca passam, mesmo que a allowlist fosse estendida.
    if (nome.startsWith("on")) continue;
    // `style` nunca passa: é um vetor de CSS injection/exfiltração
    // (`background:url(...)`, `position:fixed` para clickjacking).
    if (nome === "style") continue;
    // `src`, `srcset`, `data`, `formaction`, `xlink:href`... nunca passam.
    if (!permitidosDaTag.includes(nome) && !permitidosGlobais.includes(nome)) continue;

    if (nome === "href") {
      const seguro = hrefSeguro(valor ?? "");
      if (!seguro) continue;
      saida.push([nome, escaparAtributo(seguro)]);
      continue;
    }
    if (nome === "rel" && ehAutoFechada) continue;
    if (nome === "target") {
      // Só `_blank`; com `rel` forçado pelo chamador para noopener.
      if ((valor ?? "").toLowerCase() !== "_blank") continue;
    }
    if (valor === null) {
      // Atributo sem valor. Só se mantém se for um boolean HTML legítimo
      // da allowlist; o resto é descartado por segurança.
      if (ATRIBUTOS_FLAG_BOOLEANOS.has(nome)) saida.push([nome, null]);
      continue;
    }
    saida.push([nome, escaparAtributo(valor)]);
  }

  // Link externo com `target="_blank"` SEM `rel="noopener noreferrer"` é
  // reverse tabnabbing. Injeta o rel em vez de confiar no autor.
  if (tag === "a" && saida.some(([n]) => n === "target")) {
    const temRel = saida.some(([n, v]) => n === "rel" && (v ?? "").includes("noopener"));
    if (!temRel) {
      const idx = saida.findIndex(([n]) => n === "target");
      saida.splice(idx + 1, 0, ["rel", "noopener noreferrer nofollow"]);
    }
  }

  return saida;
}

File: backend/config/sanitizar_html.py
"""
Sanitizador de HTML editorial — fronteira primária de XSS no backend.

POR QUE EXISTE
--------------
`moderacao.PaginaEditorial.conteudo` é HTML editável pelo admin e é
consumido por `frontend/app/paginas/[slug]/page.tsx` via
`dangerouslySetInnerHTML`. Enquanto o texto não for sanitizado, qualquer
HTML com `<script>`, `onerror=` ou `javascript:` neste campo viraria XSS
armazenado servido na origem do portal.

Este é o ponto PRIMÁRIO de sanitização (roda no servidor, antes de
persistir e antes de servir pela API). O sanitizador do frontend
(`frontend/lib/sanitizar-html.ts`) é a segunda camada, útil para HTML já
gravado antes desta validação existir ou que venha de cache.

POR QUE `html.parser` E NÃO REGEX
---------------------------------
`html.parser.HTMLParser` é um tokenizador de streaming que implementa as
regras da spec HTML (raw text, atributos com/sem aspas, comentários,
CDATA,<TEntity). Regex sobre HTML é a origem clássica de bypass
(`<scr"+"ipt>`, `<img src=x onerror =alert(1)>`, comentários
condicionais). Aqui a decisão é tomada token a token, o que torna a
classe inteira de bypass de parsing estruturalmente inexistente.

POLÍTICA: ALLOWLIST
-------------------
Tags e atributos são reconstruídos a partir de uma allowlist. Tudo que
não está na allowlist é DESCARTADO. O default é "não renderizar", nunca
"não bloquear" — uma blacklist de tags perigosas é sempre incompleta
(`<svg><animate onbegin=…>`, `<math><mtext>`, `<details ontoggle>`, …).

`conteudo` de tag proibida é descartado junto (comportamento de DOMPurify
para `script`/`style`/`iframe`): manter o código-fonte de um script como
texto visível não é neutralização, é ruído.

A allowlist é deliberadamente estreita: uma página editorial (Termos,
Política de Privacidade, Sobre) é texto institucional com formatação
simples. Não há caso legítimo para `<form>`, `<iframe>`, `<object>`,
`<style>`, atributos `on*` ou `style=` inline.
"""

from __future__ import annotations

from html import escape
from html.parser import HTMLParser
from typing import Iterable

# --------------------------------------------------------------------------
# Allowlist
# --------------------------------------------------------------------------

# tag -> "fechar" (mantém o elemento) | "descartar-conteudo" (remove o miolo)
TAGS_PERMITIDAS: dict[str, str] = {
    # Estrutura e texto
    "p": "fechar",
    "br": "fechar",
    "hr": "fechar",
    "strong": "fechar",
    "b": "fechar",
    "em": "fechar",
    "i": "fechar",
    "u": "fechar",
    "s": "fechar",
    "sub": "fechar",
    "sup": "fechar",
    "small": "fechar",
    "mark": "fechar",
    "abbr": "fechar",
    "time": "fechar",
    # Títulos
    "h1": "fechar",
    "h2": "fechar",
    "h3": "fechar",
    "h4": "fechar",
    "h5": "fechar",
    "h6": "fechar",
    # Listas
    "ul": "fechar",
    "ol": "fechar",
    "li": "fechar",
    "dl": "fechar",
    "dt": "fechar",
    "dd": "fechar",
    # Citação e código
    "blockquote": "fechar",
    "pre": "fechar",
    "code": "fechar",
    "kbd": "fechar",
    "samp": "fechar",
    "var": "fechar",
    # Tabelas (usadas em tabela de preços / prazos dos Termos)
    "table": "fechar",
    "thead": "fechar",
    "tbody": "fechar",
    "tfoot": "fechar",
    "tr": "fechar",
    "th": "fechar",
    "td": "fechar",
    "caption": "fechar",
    "col": "fechar",
    "colgroup": "fechar",
    # Link
    "a": "fechar",
    # Elementos cujo CONTEÚDO é inseguro no contexto do elemento.
    "script": "descartar-conteudo",
    "style": "descartar-conteudo",
    "iframe": "descartar-conteudo",
    "frame": "descartar-conteudo",
    "frameset": "descartar-conteudo",
    "object": "descartar-conteudo",
    "embed": "descartar-conteudo",
    "applet": "descartar-conteudo",
    "template": "descartar-conteudo",
    "noscript": "descartar-conteudo",
    "svg": "descartar-conteudo",
    "math": "descartar-conteudo",
    "canvas": "descartar-conteudo",
    "audio": "descartar-conteudo",
    "video": "descartar-conteudo",
    "form": "descartar-conteudo",
    "button": "descartar-conteudo",
    "select": "descartar-conteudo",
    "textarea": "descartar-conteudo",
    "link": "descartar-conteudo",
    "meta": "descartar-conteudo",
    "base": "descartar-conteudo",
}

# Atributos aceitos por tag. `"*"` vale para todas.
ATRIBUTOS_GLOBAIS: frozenset[str] = frozenset({"title", "lang", "dir"})
ATRIBUTOS_POR_TAG: dict[str, frozenset[str]] = {
    "a": frozenset({"href", "rel", "target"}),
    "ol": frozenset({"start", "type", "reversed"}),
    "li": frozenset({"value"}),
    "td": frozenset({"colspan", "rowspan", "headers"}),
    "th": frozenset({"colspan", "rowspan", "scope", "abbr"}),
    "col": frozenset({"span"}),
    "colgroup": frozenset({"span"}),
    "time": frozenset({"datetime"}),
    "abbr": frozenset({"title"}),
}

# Atributos booleanos legítimos (presentes sem valor).
ATRIBUTOS_BOOLEANOS: frozenset[str] = frozenset({"reversed"})

# Esquemas tolerados em `href`.
ESQUEMAS_HREF: frozenset[str] = frozenset({"http", "https", "mailto"})

# `target` aceito. Qualquer outro valor (inclusive `_top`, que sobe
# para o frame superior) é descartado — não ampliamos o pedido do autor.
TARGETS_ACEITOS: frozenset[str] = frozenset({"_blank"})


def _href_seguro(valor: str) -> str | None:
    """
    Valida e canonicaliza um `href`. Devolve `None` se não for permitido.

    Não usa comparação por substring nem `startswith("https://")`: o
    esquema é extraído com `urllib.parse.urlsplit`, que faz a análise
    completa (e é o mesmo parser que o browser usa para o início da URL).
    """
    from urllib.parse import urlsplit

    bruto = (valor or "").strip()
    if not bruto:
        return None

    # Remove controles C0/DEL. Neutraliza `java\tscript:alert(1)` — o
    # browser ignora tab/CR/LF dentro de uma URL. ESPAÇO não é removido:
    # `java script:` vira esquema inválido e é recusado logo abaixo.
    bruto = "".join(c for c in bruto if ord(c) > 0x1F and ord(c) != 0x7F)
    if not bruto:
        return None

    try:
        partes = urlsplit(bruto)
    except ValueError:
        return None

    esquema = (partes.scheme or "").lower()
    if not esquema:
        # Relativa: só aceita caminho absoluto interno (`/politica`).
        # `//evil.tld` é scheme-relative e resolve para QUALQUER host.
        if bruto.startswith("/") and not bruto.startswith("//"):
            return bruto
        return None

    if esquema not in ESQUEMAS_HREF:
        return None

    # Credenciais embutidas (phishing com URL "confiável").
    if esquema != "mailto" and partes.username:
        return None
    # `mailto:` não tem host; as demais exigem.
    if esquema != "mailto" and not partes.netloc:
        return None

    return bruto


def _atributos_permitidos(tag: str) -> frozenset[str]:
    return ATRIBUTOS_GLOBAIS | ATRIBUTOS_POR_TAG.get(tag, frozenset())


class _Sanitizador(HTMLParser):
    """
    Reconstrói o documento a partir de uma allowlist.

    `HTMLParser` já trata raw text (`<script>`), comentários, CDATA e
    aspas em atributos — que é exatamente a superfície onde sanitizadores
    baseados em regex costumam ser furados.
    """

    def __init__(self) -> None:
        # `convert_charrefs=False` para não reaplicar escape em texto que
        # já passamos por `escape()`.
        super().__init__(convert_charrefs=False)
        self._saida: list[str] = []
        self._pular_ate: str | None = None
        self._pular_profundidade = 0

    # -- texto -------------------------------------------------------------
    def handle_data(self, data: str) -> None:
        if self._pular_ate is not None:
            return
        # `quote=True` escapa também `"` e `'`. Inofensivo em text node e
        # necessário se a saída um dia for embutida num atributo — mesma
        # decisão do sanitizador do frontend, pelo mesmo motivo.
        self._saida.append(escape(data, quote=True))

    def handle_entityref(self, name: str) -> None:  # ex.: `&amp;`
        if self._pular_ate is not None:
            return
        self._saida.append(f"&{name};")

    def handle_charref(self, name: str) -> None:  # ex.: `&#38;`
        if self._pular_ate is not None:
            return
        self._saida.append(f"&#{name};")

    def handle_comment(self, data: str) -> None:
        # Comentários HTML são clássico vetor de conditional comment e de
        # smuggling em parsers lenient. Descartados.
        return

    def handle_decl(self, decl: str) -> None:
        return

    def handle_pi(self, data: str) -> None:
        return

    def unknown_decl(self, data: str) -> None:
        # CDATA, por exemplo.
        return

    # -- elementos ---------------------------------------------------------
    def handle_starttag(self, tag: str, attrs: Iterable[tuple[str, str | None]]) -> None:
        nome = tag.lower()
        if self._pular_ate is not None:
            # Dentro de um elemento "descartar-conteudo": só importa
            #rementar profundidade para aninhamento da MESMA tag.
            if nome == self._pular_ate:
                self._pular_profundidade += 1
            return

        modo = TAGS_PERMITIDAS.get(nome)
        if modo is None:
            # Tag fora da allowlist: descarta a TAG e preserva o texto
            # interno (que o browser renderizaria como texto puro).
            return
        if modo == "descartar-conteudo":
            self._pular_ate = nome
            self._pular_profundidade = 1
            return

        self._saida.append(self._tag_abertura(nome, attrs))

    def handle_startendtag(self, tag: str, attrs: Iterable[tuple[str, str | None]]) -> None:
        nome = tag.lower()
        if self._pular_ate is not None:
            return
        if nome not in TAGS_PERMITIDAS or TAGS_PERMITIDAS[nome] == "descartar-conteudo":
            return
        # `convert_charrefs=False` não afeta self-closing; emite `<br/>`
        # para tags que a spec define como vazias.
        self._saida.append(self._tag_abertura(nome, attrs) + (" />" if nome in _VAZIAS else ">"))

    def handle_endtag(self, tag: str) -> None:
        nome = tag.lower()
        if self._pular_ate is not None:
            if nome == self._pular_ate:
                self._pular_profundidade -= 1
                if self._pular_profundidade <= 0:
                    self._pular_ate = None
                    self._pular_profundidade = 0
            return
        if nome in TAGS_PERMITIDAS and TAGS_PERMITIDAS[nome] == "fechar":
            self._saida.append(f"</{nome}>")

    # -- serialização ------------------------------------------------------
    def _tag_abertura(self, nome: str, attrs: Iterable[tuple[str, str | None]]) -> str:
        permitidos = _atributos_permitidos(nome)
        vistos: set[str] = set()
        pares: list[tuple[str, str | None]] = []
        for bruto_nome, valor in attrs:
            chave = (bruto_nome or "").lower()
            if chave in vistos:
                continue  # duplicata: vence a primeira (o browser também usa a 1ª)
            vistos.add(chave)
            if chave.startswith("on"):
                continue  # handler: nunca passa
            if chave in {"style", "src", "srcset", "data", "formaction", "action", "background"}:
                continue
            if chave not in permitidos:
                continue
            if chave == "href":
                seguro = _href_seguro(valor or "")
                if seguro is None:
                    continue
                pares.append((chave, seguro))
                continue
            if chave == "target":
                if (valor or "").lower() not in TARGETS_ACEITOS:
                    continue
                pares.append((chave, (valor or "").lower()))
                continue
            if valor is None:
                if chave in ATRIBUTOS_BOOLEANOS:
                    pares.append((chave, None))
                continue
            pares.append((chave, valor))

        # `target="_blank"` sem `rel=noopener` é reverse tabnabbing. Injeta.
        tem_target = any(nome_atrib == "target" for nome_atrib, _ in pares)
        tem_noopener = any(
            nome_atrib == "rel" and "noopener" in (valor or "").lower()
            for nome_atrib, valor in pares
        )
        if tem_target and not tem_noopener:
            pares.append(("rel", "noopener noreferrer nofollow"))

        if not pares:
            return f"<{nome}>"
        serializados = "".join(
            f" {chave}" if valor is None else f' {chave}="{escape(valor, quote=True)}"'
            for chave, valor in pares
        )
        return f"<{nome}{serializados}>"

    def resultado(self) -> str:
        return "".join(self._saida)


# Tags vazias na spec HTML (não devem receber tag de fechamento).
_VAZIAS: frozenset[str] = frozenset(
    {"br", "hr", "col", "wbr", "area", "base", "img", "input", "link", "meta", "source", "track"}
)


def sanitizar_html(entrada: object) -> str:
    """
    Devolve `entrada` com apenas marcação da allowlist.

    - Nunca lança. Entrada não-string (ou `None`) vira `""`.
    - Preserva o TEXTO (escapado) e remove a marcação não permitida.
    - `html.parser` tolera HTML malformado; nunca propaga exceção para o
      render da página.
    """
    if not isinstance(entrada, str) or not entrada:
        return ""
    try:
        analisador = _Sanitizador()
        analisador.feed(entrada)
        analisador.close()
        return analisador.resultado()
    except Exception:  # noqa: BLE001 — sanitização é fail-closed, nunca fail-open
        # Se o tokenizer explodir (bug do parser, entrada bizarra), o
        # fallback seguro é NÃO renderizar HTML. Texto puro seria ainda
        # melhor, mas exigiria um segundo parser; vazio é fail-closed e
        # nunca expõe marcação não sanitizada.
        return ""


def contem_html_perigoso(entrada: object) -> bool:
    """
    `True` se a sanitização REMOVEU algo — usado para emitir log de
    auditoria quando conteúdo editorial chega sujo, sem fazer a
    requisição falhar.
    """
    if not isinstance(entrada, str):
        return False
    return sanitizar_html(entrada) != entrada
